### PR TITLE
add uv_device_t as stream on windows and Linux to handle device IO

### PR DESCRIPTION
**Pull request by zhaozg**
reopen https://github.com/libuv/libuv/pull/379
older https://github.com/libuv/libuv/pull/19
oldest https://github.com/joyent/libuv/pull/1580

TODO： rewrite simple test code.

---
This change is [Reviewable](https://reviewable.io/reviews/libuv/libuv/484)

### CODE REVIEW

**Comment by creationix** [comment]
Would this also allow creating ptty sessions with libuv or does that need another primitive?  I often have need for both creating local ptty sessions and communicating over serial to embedded devices.

File: src/unix/device.c [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#include "uv.h"
+#include "internal.h"
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+
+int uv_device_init(uv_loop_t* loop,
+                   uv_device_t* device,
+                   const char*path,
```
Comment (rphillips) [inline src/unix/device.c]: space is missing

File: src/win/device.c [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+                    unsigned int nbufs,
+                    uv_write_cb cb) {
+  int result;
+  DWORD err = 0;
+  
+  if (nbufs != 1 && (nbufs != 0 || !handle)) {
+    return ERROR_NOT_SUPPORTED;
+  }
+  /* Only TCP handles are supported for sharing. */
+  if (handle && (!(handle->flags & UV_HANDLE_CONNECTION))) {
```
Comment (rphillips) [inline src/win/device.c]: seems like this comment might be wrong... should it be ```UV_HANDLE_CONNECTION```?

**Comment by rphillips** [comment]
Does this work on OSX? Would be nice if it did.

**Comment by creationix** [comment]
@zhaozg Do you have an example of opening a serial port and talking to it?  I don't know enough about this level of C interfaces to figure it out on my own.

**Comment by zhaozg** [comment]
@rphillips I without osx box, should have no result on osx.

@creationix I have example on windows to support serial (through luv), but I think it would be work on linux. https://github.com/zhaozg/luv/tree/device https://github.com/zhaozg/luvit/tree/extend/deps/serial

**Comment by zhaozg** [comment]
https://github.com/voodootikigod/node-serialport/issues/555

**Comment by dtex** [comment]
George, I would just like to make sure I understand the state of your PR. This is far out of my area of expertise so I am not qualified to look at the code and figure it out. Forgive me if I ask obvious or overly simple questions. 

This PR adds support for Windows, Linux and OSX but it has not been tested on OSX. Is that correct?

In your response above you said "should have no result on osx". I'd just like to clarify what you meant by that.

You also said "I think it would easy to work on linux(because we have ffi)". Am I correct in understanding that your pull request does not depend on FFI. You are just confident it will work because it can be done via FFI as well. Is that correct?

Thank you for your time. This PR has the potential to be a really big deal for the Nodebots community.

**Comment by zhaozg** [comment]
> This PR adds support for Windows, Linux and OSX but it has not been tested on OSX. Is that correct?
Yes, I do some test on windows and Linux, but not on others.

**Comment by batmaninpink** [comment]
Further improvements here: https://github.com/batmaninpink/libuv/commit/9274d734af986c6718837d53231c669a5ad63cee

I am using this to implement a VPN client for Windows / Linux / OS X -  which works fine with the above commit.

**Comment by saghul** [comment]
@batmaninpink Thanks for letting us know! A couple of comments: your patch was probably targeted at an older version of this patch, some of the changes no longer apply. And the `uv_device_init_fd` should be a `uv_device_open`, analogous to `uv_{tcp,udp,pipe}_open`.

**Comment by zhaozg** [comment]
Good msg from @batmaninpink I'll merge some code from your version. thanks.

**Comment by saghul** [comment]
@zhaozg I suggest you don't. See my comment here: https://github.com/libuv/libuv/pull/484#issuecomment-138868084

**Comment by zhaozg** [comment]
> @zhaozg I suggest you don't. See my comment here: #484 (comment)

I notice that, I'll add uv_device_open.

**Comment by batmaninpink** [comment]
Doh, updated here: https://github.com/batmaninpink/libuv/commit/72553a2bb01b09c0e5b66a6672380443ea5fccec

File: docs/src/device.rst [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+.. seealso:: The :c:type:`uv_stream_t` members also apply.
+
+
+API
+---
+
+.. c:function::  int uv_device_init(uv_loop_t* handle, uv_device_t* device, const char* path, int flags);
+
+    Initialize a new device with the given path. `path` must point a dev path, 
+    just ``/dev/device_name`` on linux, bsd or osx, and ``\.\Global\DeviceName``
```
Comment (saghul) [inline docs/src/device.rst]: s/just/for example/

File: docs/src/device.rst [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    or \\.\DeviceName on windows. `flags` must be O_RDONLY,O_WRONLY or O_RDWR.
```
Comment (saghul) [inline docs/src/device.rst]: Quoting on DeviceName.

Comment (saghul) [inline docs/src/device.rst]: On Unix we might need more flags for open, like O_NOCTTY when opening a serial device.

Comment (saghul) [inline docs/src/device.rst]: Space after comma.

File: docs/src/device.rst [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+.. c:function::  int uv_device_open(uv_loop_t* handle, uv_device_t* device, uv_os_fd_t fd, int flags);
+
+    Initialize a new device with the given file descriptor. `fd` must be a dev
+    file descriptor. `flags` must be O_RDONLY,O_WRONLY or O_RDWR.
```
Comment (saghul) [inline docs/src/device.rst]: I'd say drop the flags. If we are passed an fd we assume it's readable and writable, if it's not the read or write operations will fail no problem.

File: docs/src/device.rst [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+.. c:function:: int uv_device_ioctl(uv_device_t* device,
```
Comment (saghul) [inline docs/src/device.rst]: Lengthy comment on this not inline.

File: src/unix/device.c [hunk 7]
```diff
@@ -0,0 +1,97 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (saghul) [inline src/unix/device.c]: Please set the copyright to "The libuv project and contributors." Here and in other files you added.

File: src/unix/device.c [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+#include "internal.h"
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+
+int uv_device_open(uv_loop_t* loop,
+                   uv_device_t* device,
+                   uv_os_fd_t fd,
+                   int flags) {
```
Comment (saghul) [inline src/unix/device.c]: As I mentioned above, drop the flags and assume UV_STREAM_READABLE | UV_STREAM_WRITABLE.

File: src/unix/device.c [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                   uv_device_t* device,
+                   const char* path,
+                   int flags) {
+  int fd, err;
+
+  assert(device);
+  if (flags != O_RDONLY && flags != O_WRONLY && flags != O_RDWR)
+    return -EINVAL;
+
+  fd = open(path, flags);
```
Comment (saghul) [inline src/unix/device.c]: Maybe we should stat the fd to check if the fd is the right type.

File: src/unix/device.c [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  int fd, err;
+
+  assert(device);
+  if (flags != O_RDONLY && flags != O_WRONLY && flags != O_RDWR)
+    return -EINVAL;
+
+  fd = open(path, flags); 
+  if (fd < 0) {
+    return -errno;
+  }
```
Comment (saghul) [inline src/unix/device.c]: Drop braces for single statement if statements.

File: src/unix/device.c [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                    uv_ioargs_t* args) {
+  int err = ioctl(uv__stream_fd((uv_stream_t*) device), cmd, args->arg);
+  if (err < 0) {
+    return -errno;
+  }
+  return err;
+}
+
+void uv__device_close(uv_device_t* handle) {
+  uv__stream_close((uv_stream_t*)handle);
```
Comment (saghul) [inline src/unix/device.c]: space between ) and handle.

File: src/win/device.c [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+#include <fcntl.h>
+
+#include "uv.h"
+#include "internal.h"
+#include "handle-inl.h"
+#include "stream-inl.h"
+#include "req-inl.h"
+
+/* A zero-size buffer for use by uv_device_read */
+static char uv_zero_[] = "";
```
Comment (saghul) [inline src/win/device.c]: Unused.

File: src/win/device.c [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+
+int uv_device_open(uv_loop_t* loop,
+                   uv_device_t* device,
+                   uv_os_fd_t fd,
+                   int flags) {
+  int uvflags = 0;
+
+  assert(device);
+  if (flags != O_RDONLY && flags != O_WRONLY && flags != O_RDWR) {
+    return -EINVAL;
```
Comment (saghul) [inline src/win/device.c]: UV_EINVAL, on windows we don't use -errno for errors.

File: src/win/device.c [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+#include "stream-inl.h"
+#include "req-inl.h"
+
+/* A zero-size buffer for use by uv_device_read */
+static char uv_zero_[] = "";
+
+int uv_device_open(uv_loop_t* loop,
+                   uv_device_t* device,
+                   uv_os_fd_t fd,
+                   int flags) {
```
Comment (saghul) [inline src/win/device.c]: Same as above.

File: src/win/device.c [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+  device->flags |= uvflags;
+  uv_connection_init((uv_stream_t*) device);
+  return 0;
+}
+
+int uv_device_init(uv_loop_t* loop,
+                   uv_device_t* device,
+                   const char* path,
+                   int flags) {
+  HANDLE *handle;
```
Comment (saghul) [inline src/win/device.c]: HANDLE*

File: src/win/device.c [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+    err = GetLastError();
+    if (r == 0 && err == ERROR_IO_PENDING) {
+      /* The req will be processed with IOCP. */
+      handle->flags |= UV_HANDLE_READ_PENDING;
+      handle->reqs_pending++;
+    } else {
+      /* Make this req pending reporting an error. */
+      SET_REQ_ERROR(req, err);
+      uv_insert_pending_req(loop, (uv_req_t*) req);
+      handle->reqs_pending++;
```
Comment (saghul) [inline src/win/device.c]: Move this out, we do it in all code branches.

File: src/win/device.c [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+  handle->flags |= UV_HANDLE_READING;
+  INCREASE_ACTIVE_COUNT(loop, handle);
+  handle->read_cb = read_cb;
+  handle->alloc_cb = alloc_cb;
+
+  /* If reading was stopped and then started again, there could still be a */
+  /* read request pending. */
+  if (handle->flags & UV_HANDLE_READ_PENDING) {
+    return 0;
+  }
```
Comment (saghul) [inline src/win/device.c]: Drop braces.

File: src/win/device.c [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+
+  if (result) {
+    /* Request completed immediately. */
+    req->u.io.queued_bytes = 0;
+    handle->reqs_pending++;
+    handle->stream.conn.write_reqs_pending++;
+    REGISTER_HANDLE_REQ(loop, handle, req);
+    uv_insert_pending_req(loop, (uv_req_t*) req);
+  } else {
+    /* Request queued by the kernel. */
```
Comment (saghul) [inline src/win/device.c]: Move this comment down after the check.

File: src/win/device.c [hunk 19]
```diff
@@ -0,0 +1,360 @@
...
+        DECREASE_ACTIVE_COUNT(loop, handle);
+      }
+      handle->flags &= ~UV_HANDLE_READABLE;
+
+      handle->read_cb((uv_stream_t*) handle, UV_EOF, &handle->read_buffer);
+      goto done;
+    }
+  }
+  /* 
+    may be we need more do read operation
```
Comment (saghul) [inline src/win/device.c]: "maybe". Also use a single line for the comment.

File: src/win/device.c [hunk 20]
```diff
@@ -0,0 +1,360 @@
...
+
+  DECREASE_PENDING_REQ_COUNT(handle);
+}
+
+
+void uv_device_close(uv_loop_t* loop, uv_device_t* device) {
+  if (device->flags & UV_HANDLE_READING) {
+    device->flags &= ~UV_HANDLE_READING;
+    DECREASE_ACTIVE_COUNT(loop, device);
+  }
```
Comment (saghul) [inline src/win/device.c]: Do we need to cancel the pending read? (with CancelIo or something)

File: src/win/internal.h [hunk 21]
```diff
@@ -166,6 +166,25 @@ void uv_process_udp_send_req(uv_loop_t* loop, uv_udp_t* handle,
 void uv_udp_close(uv_loop_t* loop, uv_udp_t* handle);
 void uv_udp_endgame(uv_loop_t* loop, uv_udp_t* handle);
 
+/*
+ * DEVICE
```
Comment (saghul) [inline src/win/internal.h]: Device

File: src/win/internal.h [hunk 22]
```diff
@@ -166,6 +166,25 @@ void uv_process_udp_send_req(uv_loop_t* loop, uv_udp_t* handle,
...
+ */
+int uv_device_read_start(uv_device_t* handle, uv_alloc_cb alloc_cb,
```
Comment (saghul) [inline src/win/internal.h]: Put one argument per line on these definitions.

File: src/win/device.c [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+  } else if (flags & O_RDWR) {
+    dwCreationDisposition |= (GENERIC_READ | GENERIC_WRITE);
+  }
+
+  handle = CreateFile(path,
+                      dwCreationDisposition,
+                      0,
+                      NULL,
+                      OPEN_EXISTING,
+                      FILE_ATTRIBUTE_SYSTEM | FILE_FLAG_OVERLAPPED,
```
Comment (saghul) [inline src/win/device.c]: Is FILE_ATTRIBUTE_SYSTEM needed? Does it imply elevated privileges?

**Comment by saghul** [comment]
Hi!

I'm terribly sorry this has been sitting here for such a long time. I finally found the time and mindset to review it thoroughly. I left a number of comments inline, and down here I'll comment on more general terms.

Note: many people seem to want this to simplify serial port interaction, so I checked node-serialport and python-serial to see if the current design would fit all cases.

1- uv_device_ioctl needs to go

uv_device_t is inherently platform specific, and neither the Windows or Unix implementations add anything over the native functions. Since we now have `uv_fileno`, users can easily use it to get the fd or HANDLE and use native APIs to do whatever is needed.

2- tests are overly complex

AFAIS the tests need elevated privileges / root access. They are also quite convoluted due to the many ifdefs and they use system (that could be replaced by uv_spawn, but it would still require root). I think a better approach would be to use some loopback virtual serial port. I found com0com (for Windows) and tty0tty (for Linux), which IMHO are acceptable dependencies, we can gracefully fail and skip the test if the device cannot be opened.


All feedback is welcome, so everyone interested, please comment!

PS: No, we're not going to add a cross-platform serial port interface, this is a building block for it, among other things.

**Comment by tigusoft-vm** [comment]
@saghul thanks for the review.

We're working on TUN/TAP for cjdns, and would like to use @zhaozg version (thanks for developing that patch).

Please clarify, is this the best idea for now to use the API like in above patch, that is to modify our program to use uv_device, in hopes that this patch 484 or one with similar API will make it to the libuv later?

For now other option is to use @cjdelisle's patch and his API.

Or what else should we use, if we should not use either of those and only use uv_fileno, then could you point us to a best-practice example of using uv_fileno to handle read/writes to something like TUN/TAP that we should follow in our code?

File: src/win/device.c [hunk 24]
```diff
@@ -0,0 +1,360 @@
...
+  DWORD err;
+
+  assert(handle->type == UV_DEVICE);
+  assert(handle->flags & UV_HANDLE_READING);
+
+  handle->flags &= ~UV_HANDLE_READ_PENDING;
+
+  if (!REQ_SUCCESS(req)) {
+    /* An error occurred doing the read. */
+    if (handle->flags & UV_HANDLE_READING)  {
```
Comment (batmaninpink) [inline src/win/device.c]: unneeded check.

File: src/win/device.c [hunk 25]
```diff
@@ -0,0 +1,360 @@
...
+      /* Successful read */
+      handle->read_cb((uv_stream_t*) handle,
+                      req->u.io.overlapped.InternalHigh,
+                      &handle->read_buffer);
+      /* Read again only if bytes == buf.len */
+      if (req->u.io.overlapped.InternalHigh < sizeof(handle->read_buffer.len))
+        goto done;
+    } else {
+      /* Connection closed */
+      if (handle->flags & UV_HANDLE_READING) {
```
Comment (batmaninpink) [inline src/win/device.c]: unneeded check.

File: src/win/device.c [hunk 26]
```diff
@@ -0,0 +1,360 @@
...
+                      &handle->read_buffer);
+    }
+  } else {
+    if (req->u.io.overlapped.InternalHigh > 0) {
+      /* Successful read */
+      handle->read_cb((uv_stream_t*) handle,
+                      req->u.io.overlapped.InternalHigh,
+                      &handle->read_buffer);
+      /* Read again only if bytes == buf.len */
+      if (req->u.io.overlapped.InternalHigh < sizeof(handle->read_buffer.len))
```
Comment (batmaninpink) [inline src/win/device.c]: meaningless check since it just jumps to done, which will happen anyway

File: src/win/device.c [hunk 27]
```diff
@@ -0,0 +1,360 @@
...
+    } else {
+      /* Connection closed */
+      if (handle->flags & UV_HANDLE_READING) {
+        handle->flags &= ~UV_HANDLE_READING;
+        DECREASE_ACTIVE_COUNT(loop, handle);
+      }
+      handle->flags &= ~UV_HANDLE_READABLE;
+
+      handle->read_cb((uv_stream_t*) handle, UV_EOF, &handle->read_buffer);
+      goto done;
```
Comment (batmaninpink) [inline src/win/device.c]: meaningless goto

File: src/win/device.c [hunk 28]
```diff
@@ -0,0 +1,360 @@
...
+    }
+  }
+  /* 
+    may be we need more do read operation 
+  */
+done:
```
Comment (batmaninpink) [inline src/win/device.c]: not needed after removing the previous goto's

File: src/win/device.c [hunk 29]
```diff
@@ -0,0 +1,360 @@
...
+      handle->flags |= UV_HANDLE_READ_PENDING;
+      handle->reqs_pending++;
+    } else {
+      /* Make this req pending reporting an error. */
+      SET_REQ_ERROR(req, err);
+      uv_insert_pending_req(loop, (uv_req_t*) req);
+      handle->reqs_pending++;
+    }
+  } 
+};
```
Comment (batmaninpink) [inline src/win/device.c]: extra ;

File: src/win/internal.h (hunk 22 above)
Comment (zhaozg) [inline src/win/internal.h]: That keep same style function definition like other functions in same file. If changed will break style internal.h.

File: src/win/device.c (hunk 23 above)
Comment (zhaozg) [inline src/win/device.c]: According to https://msdn.microsoft.com/en-us/library/ee332330(VS.85).aspx FILE_ATTRIBUTE_SYSTEM: The file is part of or used exclusively by an operating system.

Comment (saghul) [inline src/win/device.c]: Is it needed for opening TAP devices or does it require elevated privileges?

Comment (zhaozg) [inline src/win/device.c]: I need do some test to answer

**Comment by zhaozg** [comment]
> 2- tests are overly complex

Yes, now  device tests are overly complex. But I think we should try to use the device that not need to install. At least TAP available on linux.

We need to find more test way for device. E.g. we can get hard disk SN on windows and ptty on linux.

**Comment by mikeal** [comment]
> We need to find more test way for device.

We have a bunch of devices in the nodejs build and test infrastructure. In the future we could incorporate some integration tests there.

File: src/win/device.c (hunk 23 above)
Comment (zhaozg) [inline src/win/device.c]: When I remove flag FILE_ATTRIBUTE_SYSTEM, Windows TAP device works well. But that is not enough reason to remove it.

**Comment by zhaozg** [comment]
To run device test, you need administrator/root privileges, and on windows need

1. install tap-windows, http://build.openvpn.net/downloads/releases/tap-windows-9.21.0.exe .
2. your windows firewall permit ICMP ping pass through.

**Comment by saghul** [comment]
@tigusoft-vm hard to tell. I like this approach better, and if the suggeste dchanges are made we should be able to merge it to libuv v1.x so no patched version is necessary. Alas there is no ETA for this.

**Comment by saghul** [comment]
@zhaozg Thanks! I'll review your newest patches as soon as I can.

**Comment by saghul** [comment]
Ok, review time, part 2.

While I was lost somewhere in south america a month ago (somethimes you have to go far far away to get some perspective), it occurred to me we can simplify this even further.

The more I look at this code the more it looks to me that what we really want is something like `uv_poll_t` but with `uv_stream_t` semantics. So how about we do that? Since this is a helper for platform-dependent code, the user is responsible for opening the fd as they please (see my previous comments about the flags to open) and then it can be used with `uv_stream_t`-like semantics (read and write only, that is).

Here is the only new API we would need (I'm not thrilled about the name, but can't come up with a better name right now)

~~~~
UV_EXTERN int uv_fdstream_init(uv_loop_t* loop,
                                                          uv_fdstream_t* fdstream,
                                                          uv_os_fd_t fd);
~~~~

Then we can modify the stream internals to give proper errors for connect/shutdown/accept.

Note that the `fd` is given as a `uv_os_fd_t`, which on Unix is just an `int`, so this will work with any fd epoll and friends can work with, while on Windows it's a `HANDLE`, which would be used with `ReadFile` and friends.

IMHO this gives us the flexibility we need, but before we move forward I'd like to hear from @bnoordhuis and/or @piscisaureus. Folks, no need to read the whole thing (nor the code) just quickly review the proposal in this comment.

If we go this route, the tests can also be greatly simplified by using a socket / pipe as the fd in order to test the whole machinery, so bonus nachos!

**Comment by bnoordhuis** [comment]
@saghul How is `uv_fdstream_t` different from `uv_stream_t`?  Put another way, how is `uv_fdstream_init()` different from a hypothetical `uv_stream_open()`?

**Comment by saghul** [comment]
@bnoordhuis Good question :-) `uv_fdstream_init` wouldn't be different from a hypothetical `uv_stream_open`, but then we wouldn't have the ability to give an error for functions which do not work over this type of stream (like `uv_write2` or `uv_shutdown`). So far `uv_stream_t`, is kind of an abstract type, we always create specific instances of streams, hence the new type.

**Comment by bnoordhuis** [comment]
Makes sense.  Seems like a good approach.

**Comment by saghul** [comment]
Thanks Ben, I think we're finally close to having something we can work with here!

@zhaozg please take a look at https://github.com/libuv/libuv/pull/484#issuecomment-157315601

**Comment by creationix** [comment]
How would this be different from the `uv_pipe_open` trick I use at https://luvit.io/blog/pty-ffi.html?

Also, while I think it's fine to limit libuv's scope to not handle the nasty platform specific details for initializing the fd, many people will be disappointed that there is no library for creating serial or pty streams cross-platform.

It would be awesome if this code was written once in C as a libuv addon library that the various libuv bindings can then use instead of each platform implementing it on their own and tightly coupling to their system.

Just like I can use libuv without having to touch node.js's source code, I'd love to use the logic in node-serialport without messing with any V8 bindings.

Should the libuv project do anything to encourage, organize, or even vet such libraries?

**Comment by saghul** [comment]
> How would this be different from the uv_pipe_open trick I use at https://luvit.io/blog/pty-ffi.html?
> 

It's almost the same, except it's not a pipe, just a fd. So no binding, connecting, etc. This new stream type would be a better fit for that hack, except for the fact that we are fixing `uv_tty_t` to support ptys in #528.

> Also, while I think it's fine to limit libuv's scope to not handle the nasty platform specific details for initializing the fd, many people will be disappointed that there is no library for creating serial or pty streams cross-platform.
> 

This is a building block, that can be built on top, either as part of libuv or as a separate project, but we are not there yet. Also, nobody proposed a libuv-level API for serial ports yet, but generally speaking I'd rather see it as a separate project.

> It would be awesome if this code was written once in C as a libuv addon library that the various libuv bindings can then use instead of each platform implementing it on their own and tightly coupling to their system.
> 
> Just like I can use libuv without having to touch node.js's source code, I'd love to use the logic in node-serialport without messing with any V8 bindings.
> 

That would indeed be nice. It could be some "libuv-serial" thing which lives in the libuv GH org if someone volunteers to write and maintain the code, I guess.

> Should the libuv project do anything to encourage, organize, or even vet such libraries?

(Personally) I only have so much time, and I don't see libuv as a [UBA](https://www.youtube.com/watch?v=lfbhiCXNPJk), so I'd rather see that as a separate library which links with libuv, since it wouldn't use any of its internals. If such a project is created I'd give it my blessing for adding it to the libuv organizartion.

**Comment by creationix** [comment]
@saghul thanks, sounds good.  Now to see if I can find such a maintainer.

**Comment by dtex** [comment]
@saghul So libuv-serial (or whatever it will be called) would not be dependency of libuv and would have to be added as a separate dependency for any libuv dependent that wanted to open a serialport correct?

**Comment by cjdelisle** [comment]
Hmm I seem to recall the pipe trick does *not* work on windows, it works well on linux/max/bsd (and I use it), I don't remember exactly why it doesn't work on windows because the patch is about 2 years old now.

**Comment by saghul** [comment]
> So libuv-serial (or whatever it will be called) would not be dependency of libuv and would have to be added as a separate dependency for any libuv dependent that wanted to open a serialport correct?

No, the project would most likely depend on libuv, since it would use the new `uv_fdstream_t` to do the reads and writes.

**Comment by cjdelisle** [comment]
Ok now I remember, uv_pipe on windows does a read with a 0 byte buffer (or maybe 1?) which is effectively a way to poll, when you do this with OpenVPN TAP, it will call you when it has a packet and because your buffer is smaller than the packet, it will trash the packet and "tough luck, should have used 1500"...   This is why I had to write the patch in the beginning.

**Comment by saghul** [comment]
Thanks for the reminder @cjdelisle!

**Comment by dtex** [comment]
Is it possible that the methods exposed by this approach will be sufficient for us to do all the platform specific machinations for opening a serialport in javascript?

**Comment by saghul** [comment]
I assume yes, but since that project doesn't exist I can't really say.

PS: Please let's keep this discussion about this PR, feel free to take it to the mailing list.

**Comment by zhaozg** [comment]
@saghul Thank you for your point of view and agree use more flexible abstract type to replace device, and you make more near to our aims.
In my opinion, should make a general API to achieve ReadFile/WriteFIle on windows and read/write on other OS for  uv_os_fd_t (not socket),  which can maybe stream or not stream, as long as event notify, just like tcp or udp on network, that feature will make IoT(Internet of Things) more efficient.
And whether pipe,device,tty,fs .. can build up on common layer?
Finally, I hope libuv team do more work on this field.

**Comment by saghul** [comment]
@zhaozg Not sure I understood you. The idea here would be to create a **new** type of stream, which would do exactly that: read/write on Unix and ReadFile/WriteFile on Windows. Most of the `uv_stream_t` internals can be used, since this would be a type of stream.

As for building tcp, pipe or tty on top of it, I don't think it helps in any way, they would just be siblings.

**Comment by zhaozg** [comment]
sorry for my poor English, I try give tell my opinion clear and simply.

1   IO operation can be split to socket(SOCKET) operation and file(uv_os_fd_t) operation.
2.  file(uv_os_fd_t) operation support pipe, tty,device,..., all use ReadFile/WriteFile and read/write apis, but different init process, and maybe have special control api.
3.  serial and IoT support should easy setup based on file operation.
4.  socket operation not need change, but file operation maybe need do refactor.

**Comment by saghul** [comment]
No worries! I know see what you mean. However, right now while pipe and tty use the ReadFile / WriteFile API, they do the 0 reads which we cannot do, see [this comment](https://github.com/libuv/libuv/pull/484#issuecomment-157409476).

IMHO, I'd modify the current PR to match the agreed upon API, and once ready to review, we can see if / where there is code duplication, and refactor as necessary.

**Comment by tigusoft-vm** [comment]
@zhaozg @saghul Thanks for your effort on this topic.

Currently we try to use this new uv_device_t APIs for cjdns project.

We use it's functions ->read_cb() ->alloc_cb() and finally we do WinAPI call:
ReadFile( handle->handle  ... ,   handle->read_req->u.io.overlapped
where handle is uv_device_t.

Will this parts of the API stay the same probably?

We seem to do fine while using winapi like that directly (it's all in ifdefs anyway) though maybe we should later check for a way to do that part too on higher level provided by uv_device_t.

**Comment by saghul** [comment]
@tigusoft-vm The overlapped structure is not part of the public API and won't be. The read data will come in the read_cb. That is what the current patch does, and there are no plans to change that.

**Comment by da2ce7** [comment]
@zhaozg thankyou for the rebase :+1:

**Comment by TomCrypto** [comment]
What's the status on this? We need uv_device_t to interact with a virtual network interface driven by a NDIS kernel driver and there is effectively no way to efficiently do it through the libuv API; the only way is to periodically poll using uv_idle_t, which is fairly nasty.

**Comment by saghul** [comment]
I guess / hope @zhaozg is working on it, and will be ready when it's ready :-)

**Comment by dtex** [comment]
HI @saghul, Is this comment:

> IMHO, I'd modify the current PR to match the agreed upon API, and once ready to review, we can see if / where there is code duplication, and refactor as necessary.

still the current state of things? (plus a rebase of course)

**Comment by saghul** [comment]
@dtex The current state does not match my proposal [here](https://github.com/libuv/libuv/pull/484#issuecomment-157315601) which is what was agreed upon last time. Last update was almost a year ago, so I guess this one is stalled.

**Comment by dtex** [comment]
HI @zhaozg Are you still interested in this pull request? There are a few of us who are looking forward to seeing some variation of it land. If you are no longer interested, that is understandable. We will try to find someone to pick it up and try to match the proposal from @saghul.

**Comment by zhaozg** [comment]
hi @dtex I'm keeping interesting on it, but I have no time to finish, So please fell free to do any change on.

**Comment by tigusoft-vm** [comment]
@dtex @zhaozg  if no one else steps up, then we can pick it up in this week.
Either way we can confirm usage of this patch as the TUN interface in (our fork of) Cjdns.

**Comment by creationix** [comment]
I'm still very interested from the point of view of serial ports for robotics projects.  I don't know how much I can help, but am willing to try.

**Comment by rvagg** [comment]
ping

All interested here: you may be best to take this code and create a new PR ontop of it to continue progress, best not block on @zhaozg being able to find time.

**Comment by tigusoft** [comment]
(btw. we are not developing this code now after all, but if there is something to be tested then ping me) thanks for work on this PR.

**Comment by nomonstershit** [comment]
will there be any support for this ? It has been a year, assuming it is about serial port. I somewhow need serial port support on libuv.

**Comment by winterland1989** [comment]
Just happened to come across this issue, while i think `uv_fdstream_t` is a good idea, i guess we can do better, how about add a `is_socket` field to `uv_stream_t` and do something special on windows? Then we can keep API the same.

BTW: this is the way [how haskell do it](http://hackage.haskell.org/package/base-4.9.1.0/docs/src/GHC.IO.FD.html#line-608).

**Comment by cjdelisle** [comment]
I cannot believe that @zhaozg has been continuously trying to get this PR merged since 2014. @zhaozg you are a hero.

I also can't believe that someone can try for 6 years to get a simple problem solved in Libuv and even after pages and pages of commits and re-bases, it seems no closer than it ever was.

I know the rumors about intrigue behind the io.js schism which ended in Joyent losing control over node - and I'm as concerned as anybody about powerful companies seizing effective control over the direction of open source projects - but if contributors who "play by the rules" risk ending up in a *code-review tarpit* then we must accept the fact that businesses will not trust OSS projects they don't control, and "hostile takeovers" will only increase in the future.

For myself, I'm glad I didn't invest any significant effort in trying to upstream my patch to solve this problem, but my fork is also over 6 years old and I need to upgrade it. However, I'm planning to deprecate Libuv entirely and slowly begin migrating to Rust where I think it would be unlikely to have a PR sit for 6 years without the maintainers making any effort to help make it acceptable.

**Comment by vtjnash** [comment]
I think that's a bit aggressive to delete that comment from @cjdelisle. It's a bit off topic, and seems to have a naive-but-understandable misunderstanding of the situation and relevant politics, so it's not particularly productive, but there's a separate button to mark and hide it as such.

**Comment by zhaozg** [comment]
closed, here https://github.com/zhaozg/libuv

**Comment by calvin2021y** [comment]
Hi @zhaozg, thanks for the patch.

**Comment by calvin2021y** [comment]
> 

no uv__process_device_shutdown_req, uv_stream_init, uv_connection_init, uv_insert_pending_req,  uv_want_endgame 
 symbol for window with your patch.

**Pull request closed without merging**